resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -986,25 +1002,34 @@ protected void doResolveAll(String inetHost,
             return;
         }
 
-        if (!doResolveAllCached(hostname, additionals, promise, resolveCache, resolvedInternetProtocolFamilies)) {
-            doResolveAllUncached(hostname, additionals, promise, promise,
-                                 resolveCache, completeOncePreferredResolved);
+        CacheResult res = doResolveAllCached(hostname, additionals, promise, resolveCache,
+            resolvedInternetProtocolFamilies);
+        if (res == CacheResult.CACHED) {
+            return;
+        }
+
+        if (res == CacheResult.CACHED_EXPIRED) {
+            promise = executor().newPromise();
         }
+
+        doResolveAllUncached(hostname, additionals, promise, promise, resolveCache, completeOncePreferredResolved);
     }
 
-    static boolean doResolveAllCached(String hostname,
+    static CacheResult doResolveAllCached(String hostname,
                                       DnsRecord[] additionals,
                                       Promise<List<InetAddress>> promise,
                                       DnsCache resolveCache,
                                       InternetProtocolFamily[] resolvedInternetProtocolFamilies) {
         final List<? extends DnsCacheEntry> cachedEntries = resolveCache.get(hostname, additionals);
         if (cachedEntries == null || cachedEntries.isEmpty()) {
-            return false;
+            return CacheResult.NOT_CACHED;
         }
 
         Throwable cause = cachedEntries.get(0).cause();
         if (cause == null) {
             List<InetAddress> result = null;
+            boolean oneEntryIsExpired = false;
+            long now = System.currentTimeMillis();
```

Comment:
should we only do this in this in a lazy fashion if we really need `now`. Also please see the comments about `System.nanoTime()`